lib/expand/join.go

```diff
@@ -744,147 +733,186 @@ func (p *Peer) stop() error {
 	return nil
 }
 
-func (p *Peer) tryConnect(operationID string) (op *operationContext, err error) {
+func (p *Peer) tryConnect(operationID string) (ctx *operationContext, err error) {
 	p.printStep("Connecting to cluster")
 	for _, addr := range p.Peers {
 		p.WithField("peer", addr).Debug("Dialing peer.")
-		op, err = p.dialWizard(addr)
+		ctx, err = p.dialWizard(addr)
 		if err == nil {
-			p.WithField("addr", op.Peer).Debug("Connected to wizard.")
+			p.WithField("addr", ctx.Peer).Debug("Connected to wizard.")
 			p.printStep("Connected to installer at %v", addr)
-			return op, nil
+			return ctx, nil
 		}
 		p.WithError(err).Info("Failed connecting to wizard.")
 		if isTerminalError(err) {
-			return nil, trace.Wrap(err)
+			return nil, utils.Abort(err)
 		}
 		// already exists error is returned when there's an ongoing install
 		// operation, do not attempt to dial the cluster until it completes
 		if trace.IsAlreadyExists(err) {
 			p.printStep("Waiting for the install operation to finish")
 			return nil, trace.Wrap(err)
 		}
-		op, err = p.dialCluster(addr, operationID)
+		ctx, err = p.dialCluster(addr, operationID)
 		if err == nil {
-			p.WithField("addr", op.Peer).Debug("Connected to cluster.")
+			p.WithField("addr", ctx.Peer).Debug("Connected to cluster.")
 			p.printStep("Connected to existing cluster at %v", addr)
-			return op, nil
+			return ctx, nil
 		}
 		p.WithError(err).Info("Failed connecting to cluster.")
-		if utils.IsAbortError(err) {
-			return nil, trace.Wrap(err)
+		if isTerminalError(err) {
+			return nil, utils.Abort(err)
 		}
 		if trace.IsCompareFailed(err) {
 			p.printStep("Waiting for another operation to finish at %v", addr)
 		}
 	}
-	return op, trace.Wrap(err)
+	return ctx, trace.Wrap(err)
 }
 
-// startAgent starts a new RPC agent using the specified operation context.
-// The agent will signal p.errC once it has terminated
-func (p *Peer) startAgent(ctx operationContext) error {
-	agent, err := p.newAgent(ctx)
+func (p *Peer) run(ctx operationContext) (dispatcher.Status, error) {
+	if ctx.Operation.Type == ops.OperationInstall {
+		return p.runAgent(ctx)
+	}
+	err := p.executeExpandOperation(ctx)
+	if err != nil {
+		return dispatcher.StatusUnknown, trace.Wrap(err)
+	}
+	return dispatcher.StatusCompleted, nil
+}
+
+func (p *Peer) runAgent(ctx operationContext) (dispatcher.Status, error) {
```

Comment:
Nit: I may be misunderstanding something, but looking at what this function does, it's not related to the "agent" in any way so the name is misleading.

Reply:
It implements the loop main loop of the joining agent during install - hence the `agent` in the name.